src/passes/Print.cpp

```diff
@@ -2560,10 +2565,17 @@ struct PrintSExpression : public UnifiedExpressionVisitor<PrintSExpression> {
   }
 
   void printDebugLocation(const Function::DebugLocation& location) {
-    if (lastPrintedLocation == location) {
+    // ;; debug location
+    // (parent
+    //  (child) ;; no debug location
+    // )
+    // ;; debug location
+    // (siblings)
+    if (lastPrintedLocation == location && indent > lastPrintedIndent) {
```

Comment:
@kripken 
Hi Alon,
the easiest way to check the debug info is remove this filter for debugging, here is my step:
1, remove this condition return 
2, build binaryen
3, run cmd: `./build/bin/wasm-opt ./test/fib-dbg.wasm -ism ./test/fib-dbg.wasm.map --print`
then wasm-opt will print all debug info.
so you can check if the new added annotation of debug info is reasonable or not.
if any other question, please feel free to ask, thank you.

Reply:
@kripken 
Hi Alon,
here is the wat file(fib-dbg.wasm.fromBinary) will full debug info printed:
```WAT
(module
 (type $i32_=>_i32 (func (param i32) (result i32)))
 (type $i32_i32_=>_none (func (param i32 i32)))
 (type $none_=>_i32 (func (result i32)))
 (type $i32_=>_none (func (param i32)))
 (type $none_=>_none (func))
 (import "env" "memory" (memory $mimport$0 256 256))
 (import "env" "table" (table $timport$0 0 0 funcref))
 (import "env" "DYNAMICTOP_PTR" (global $gimport$0 i32))
 (import "env" "tempDoublePtr" (global $gimport$1 i32))
 (import "env" "ABORT" (global $gimport$2 i32))
 (import "env" "STACKTOP" (global $gimport$3 i32))
 (import "env" "STACK_MAX" (global $gimport$4 i32))
 (import "env" "gb" (global $gimport$5 i32))
 (import "env" "fb" (global $gimport$6 i32))
 (import "global" "NaN" (global $gimport$7 f64))
 (import "global" "Infinity" (global $gimport$8 f64))
 (import "env" "memoryBase" (global $gimport$9 i32))
 (import "env" "tableBase" (global $gimport$10 i32))
 (global $global$0 (mut i32) (global.get $gimport$0))
 (global $global$1 (mut i32) (global.get $gimport$1))
 (global $global$2 (mut i32) (global.get $gimport$2))
 (global $global$3 (mut i32) (global.get $gimport$3))
 (global $global$4 (mut i32) (global.get $gimport$4))
 (global $global$5 (mut i32) (global.get $gimport$5))
 (global $global$6 (mut i32) (global.get $gimport$6))
 (global $global$7 (mut i32) (i32.const 0))
 (global $global$8 (mut i32) (i32.const 0))
 (global $global$9 (mut i32) (i32.const 0))
 (global $global$10 (mut i32) (i32.const 0))
 (global $global$11 (mut f64) (global.get $gimport$7))
 (global $global$12 (mut f64) (global.get $gimport$8))
 (global $global$13 (mut i32) (i32.const 0))
 (global $global$14 (mut i32) (i32.const 0))
 (global $global$15 (mut i32) (i32.const 0))
 (global $global$16 (mut i32) (i32.const 0))
 (global $global$17 (mut f64) (f64.const 0))
 (global $global$18 (mut i32) (i32.const 0))
 (global $global$19 (mut i32) (i32.const 0))
 (global $global$20 (mut i32) (i32.const 0))
 (global $global$21 (mut f64) (f64.const 0))
 (global $global$22 (mut i32) (i32.const 0))
 (global $global$23 (mut f64) (f64.const 0))
 (export "setThrew" (func $setThrew))
 (export "runPostSets" (func $runPostSets))
 (export "establishStackSpace" (func $establishStackSpace))
 (export "stackSave" (func $stackSave))
 (export "stackRestore" (func $stackRestore))
 (export "_fib" (func $_fib))
 (export "stackAlloc" (func $stackAlloc))
 (func $stackAlloc (param $0 i32) (result i32)
  (local $1 i32)
  (block $label$1
   (local.set $1
    (global.get $global$3)
   )
   (global.set $global$3
    (i32.add
     (global.get $global$3)
     (local.get $0)
    )
   )
   (global.set $global$3
    (i32.and
     (i32.add
      (global.get $global$3)
      (i32.const 15)
     )
     (i32.const -16)
    )
   )
   (return
    (local.get $1)
   )
  )
 )
 (func $stackSave (result i32)
  (return
   (global.get $global$3)
  )
 )
 (func $stackRestore (param $0 i32)
  (global.set $global$3
   (local.get $0)
  )
 )
 (func $establishStackSpace (param $0 i32) (param $1 i32)
  (block $label$1
   (global.set $global$3
    (local.get $0)
   )
   (global.set $global$4
    (local.get $1)
   )
  )
 )
 (func $setThrew (param $0 i32) (param $1 i32)
  (if
   (i32.eq
    (global.get $global$7)
    (i32.const 0)
   )
   (block
    (global.set $global$7
     (local.get $0)
    )
    (global.set $global$8
     (local.get $1)
    )
   )
  )
 )
 (func $_fib (param $0 i32) (result i32)
  (local $1 i32)
  (local $2 i32)
  (local $3 i32)
  (local $4 i32)
  (local $5 i32)
  (local $6 i32)
  (local $7 i32)
  (local $8 i32)
  (local $9 i32)
  (local $10 i32)
  (local $11 i32)
  (block $label$1
   (local.set $11
    (global.get $global$3)
   )
   ;;@ fib.c:3:0
   (local.set $6
    ;;@ fib.c:3:0
    (i32.gt_s
     ;;@ fib.c:3:0
     (local.get $0)
     ;;@ fib.c:3:0
     (i32.const 0)
    )
   )
   ;;@ fib.c:3:0
   (if
    ;;@ fib.c:3:0
    (local.get $6)
    (block
     ;;@ fib.c:3:0
     (local.set $1
      ;;@ fib.c:3:0
      (i32.const 0)
     )
     ;;@ fib.c:3:0
     (local.set $5
      ;;@ fib.c:3:0
      (i32.const 1)
     )
     ;;@ fib.c:3:0
     (local.set $8
      ;;@ fib.c:3:0
      (i32.const 0)
     )
    )
    (block
     ;;@ fib.c:3:0
     (local.set $4
      ;;@ fib.c:3:0
      (i32.const 1)
     )
     ;;@ fib.c:8:0
     (return
      ;;@ fib.c:8:0
      (local.get $4)
     )
    )
   )
   ;;@ fib.c:8:0
   (loop $label$4
    ;;@ fib.c:8:0
    (block $label$5
     ;;@ fib.c:4:0
     (local.set $3
      ;;@ fib.c:4:0
      (i32.add
       ;;@ fib.c:4:0
       (local.get $5)
       ;;@ fib.c:4:0
       (local.get $1)
      )
     )
     ;;@ fib.c:3:0
     (local.set $9
      ;;@ fib.c:3:0
      (i32.add
       ;;@ fib.c:3:0
       (local.get $8)
       ;;@ fib.c:3:0
       (i32.const 1)
      )
     )
     ;;@ fib.c:3:0
     (local.set $7
      ;;@ fib.c:3:0
      (i32.eq
       ;;@ fib.c:3:0
       (local.get $9)
       ;;@ fib.c:3:0
       (local.get $0)
      )
     )
     ;;@ fib.c:3:0
     (if
      ;;@ fib.c:3:0
      (local.get $7)
      (block
       ;;@ fib.c:3:0
       (local.set $4
        ;;@ fib.c:3:0
        (local.get $3)
       )
       ;;@ fib.c:3:0
       (br $label$5)
      )
      (block
       ;;@ fib.c:3:0
       (local.set $2
        ;;@ fib.c:3:0
        (local.get $5)
       )
       ;;@ fib.c:3:0
       (local.set $5
        ;;@ fib.c:3:0
        (local.get $3)
       )
       ;;@ fib.c:3:0
       (local.set $8
        ;;@ fib.c:3:0
        (local.get $9)
       )
       ;;@ fib.c:3:0
       (local.set $1
        ;;@ fib.c:3:0
        (local.get $2)
       )
      )
     )
     ;;@ fib.c:3:0
     (br $label$4)
    )
   )
   (return
    ;;@ fib.c:8:0
    (local.get $4)
   )
  )
 )
 (func $runPostSets
  (local $0 i32)
  (nop)
 )
 ;; custom section "sourceMappingURL", size 35, contents: "\"http://localhost:8000/fib.wasm.map"
)

```

Reply:
I see, thanks, and sorry for the delay.

I'll look at this again soon, but first, I have recently been debugging another source maps issue (a user reported the wrong annotation somewhere), and so I read the source maps spec again and then our code. I have to say I am still a little confused on it. It seems like the spec says nothing about ranges or how to infer debug locations for things without an annotation, but Binaryen does do that - it will keep using a debug location after it sees one, until the next. I am not sure if that is correct or not, and I'm not sure if that is related to this PR or not, but the binary format parts seem like they might be? I am quite worried about changing it, because of how it might affect other users...

For this PR specifically, since I would like to move forward on it: Is it possible to split the Print parts from the binary format parts? I think there is a low bar to landing a Print improvement, but a high bar for a binary format change (again, because users could be affected by the latter). So if we can split it, that would make review simpler, and also allow the Print part to land quickly I hope.

Reply:
@kripken 
yep, I will splite the PR and suppose the print PR will comming soon